### ISSUE TITLE
Untrusted desktop entry file (escaping issue)

### ISSUE THREAD

**Issue by develar** [TheAssassin/AppImageLauncher#170]
Version: v1.2.2

<img width="935" alt="Screenshot 2019-06-22 at 06 36 56" src="https://user-images.githubusercontent.com/350686/59959477-5a37a400-94b8-11e9-86da-1c7ba9b0cd51.png">

Linux 4.18.0-22-generic #23~18.04.1-Ubuntu SMP Thu Jun 6 08:37:25 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

Just run AppImage, accept to integrate it, and... no launcher in menu.

**Comment by TheAssassin** [comment]
That's quite odd, first time I see it. How can we make it "trusted"? Hasn't it been marked as executable?

**Comment by develar** [comment]
Here I am just a user, I didn't investigated what's "untrusted" means. The same effect was for electron-builder desktop integration script. But I didn't fix electron-builder version, because decided to use AppImageLauncher instead (this version of electron-builder not yet released).

It is ubuntu linux 18.0.4 (plus all updates `apt-get update && apt-get upgrade`) installed as VM for Parallel Desktop for macOS.

> Hasn't it been marked as executable?

Access: (0755/-rwxr-xr-x)  Uid: ( 1000/ develar)   Gid: ( 1000/ develar). So, it is executable.
<img width="641" alt="Screenshot 2019-06-22 at 15 15 23" src="https://user-images.githubusercontent.com/350686/59964416-97c01f80-9500-11e9-9d81-4e79d02e6d2d.png">

**Comment by develar** [comment]
Hmm.... I cannot reproduce for https://github.com/AkashaProject/Community/releases/download/0.7.2/AKASHA-0.7.2-x86_64.AppImage, maybe something wrong with AppImage generated by electron-builder.

**Comment by TheAssassin** [comment]
Can you post a link to your "broken" AppImage? I'll have a look.

By the way, talking bout new projects, check out https://github.com/TheAssassin/appimagelint/

**Comment by develar** [comment]
https://drive.google.com/file/d/1-dQ35KD4j61nP0Lr3iNRvx65WNaj2P67/view?usp=sharing

**Comment by develar** [comment]
```
appimagelint.cli[13313] [INFO] Running check "GNU libc ABI check"
appimagelint.glibc_abi_check[13313] [INFO] detected required version for runtime: 2.3.3
appimagelint.glibc_abi_check[13313] [INFO] detected required version for payload: 2.17
appimagelint.glibc_abi_check[13313] [INFO] [✔] AppImage can run on Debian oldstable (jessie)
appimagelint.glibc_abi_check[13313] [INFO] [✔] AppImage can run on Debian stable (stretch)
appimagelint.glibc_abi_check[13313] [INFO] [✔] AppImage can run on Debian testing (buster)
appimagelint.glibc_abi_check[13313] [INFO] [✔] AppImage can run on Debian unstable (sid)
appimagelint.glibc_abi_check[13313] [INFO] [✔] AppImage can run on Ubuntu trusty
appimagelint.glibc_abi_check[13313] [INFO] [✔] AppImage can run on Ubuntu xenial
appimagelint.glibc_abi_check[13313] [INFO] [✔] AppImage can run on Ubuntu bionic
appimagelint.glibc_abi_check[13313] [INFO] [✔] AppImage can run on Ubuntu cosmic
appimagelint.glibc_abi_check[13313] [INFO] [✔] AppImage can run on Ubuntu disco
appimagelint.cli[13313] [INFO] Running check "GNU libstdc++ ABI check"
appimagelint.glibcxx_abi_check[13313] [INFO] detected required version for runtime: <none>
appimagelint.glibcxx_abi_check[13313] [WARNING] could not find any dependencies, skipping check
appimagelint.cli[13313] [INFO] Running check "Icons validity and location check"
appimagelint.icon_check[13313] [INFO] Extracting icon name from desktop file: /tmp/.mount_electr3Ksnjt/electron-webpack-quick-start.desktop
appimagelint.icon_check[13313] [INFO] Checking resolution of icon: /tmp/.mount_electr3Ksnjt/electron-webpack-quick-start.png
appimagelint.icon_check[13313] [INFO] [✔] Valid icon in AppDir root
appimagelint.icon_check[13313] [INFO] Checking resolution of icon: /tmp/.mount_electr3Ksnjt/.DirIcon
appimagelint.icon_check[13313] [INFO] [✔] Valid icon file in .DirIcon
appimagelint.icon_check[13313] [INFO] Checking resolution of icon: /tmp/.mount_electr3Ksnjt/usr/share/icons/hicolor/128x128/apps/electron-webpack-quick-start.png
appimagelint.icon_check[13313] [INFO] Checking resolution of icon: /tmp/.mount_electr3Ksnjt/usr/share/icons/hicolor/16x16/apps/electron-webpack-quick-start.png
appimagelint.icon_check[13313] [INFO] Checking resolution of icon: /tmp/.mount_electr3Ksnjt/usr/share/icons/hicolor/256x256/apps/electron-webpack-quick-start.png
appimagelint.icon_check[13313] [INFO] Checking resolution of icon: /tmp/.mount_electr3Ksnjt/usr/share/icons/hicolor/32x32/apps/electron-webpack-quick-start.png
appimagelint.icon_check[13313] [INFO] Checking resolution of icon: /tmp/.mount_electr3Ksnjt/usr/share/icons/hicolor/48x48/apps/electron-webpack-quick-start.png
appimagelint.icon_check[13313] [INFO] Checking resolution of icon: /tmp/.mount_electr3Ksnjt/usr/share/icons/hicolor/64x64/apps/electron-webpack-quick-start.png
appimagelint.icon_check[13313] [INFO] [✔] Other integration icons valid
```

Everything is ok except `WARNING] could not find any dependencies, skipping check` and icons for 96/24 sizes (removed it, looks like it is not needed).

**Comment by TheAssassin** [comment]
Good. There's no desktop file check yet, unfortunately, but I will add one soon-ish. That'll be quite useful to you, since you don't use appimagetool, which comes with some checks.

**Comment by TheAssassin** [comment]
According to some StackOverflow issues, this dialog goes away when a desktop file is marked executable. That's what AIL is supposed to do. I will test this nevertheless on Ubuntu 18.04.
The desktop file in your AppImage looked good.

**Comment by develar** [comment]
@TheAssassin Still, something is wrong or in electron-builder AppImage, or in AppImageLauncher, because I don't see "Open with AppImageLauncher" in context menu. I will investigate next week.

<img width="363" alt="Screenshot 2019-06-23 at 09 23 01" src="https://user-images.githubusercontent.com/350686/59972965-993d2680-9598-11e9-86f6-7e87521e508d.png">
<img width="384" alt="Screenshot 2019-06-23 at 09 23 08" src="https://user-images.githubusercontent.com/350686/59972966-993d2680-9598-11e9-80e3-184073b17850.png">

**Comment by TheAssassin** [comment]
Strange, seems right over here:

```
> xxd electron-webpack-quick-start-0.0.0_00000000000000000000000000000000.AppImage | head -n1
00000000: 7f45 4c46 0201 0100 4149 0200 0000 0000  .ELF....AI......
> xdg-mime query filetype electron-webpack-quick-start-0.0.0_00000000000000000000000000000000.AppImage 
application/vnd.appimage
```

Can you please run the second command and check whether AIL's MIME type definitions at least work? Might be a bug in Ubuntu's desktop, and we need to update some caches or so.

**Comment by develar** [comment]
```
develar@develar-ubuntu:~$ xdg-mime query filetype ~/Desktop/electron-webpack-quick-start\ 0.0.0.AppImage 
application/vnd.appimage
```

Reported correctly, but still no option in menu :)

**Comment by develar** [comment]
There is an option in context menu if I download via firefox.

Still, app is not added to menu. And in the generated desktop file I see 

```
TryExec=/home/develar/Applications/electron-webpack-quick-start 0.0.0 (copy)_00000000000000000000000000000000.AppImage
```

but from what I know, spaces for TryExec must be escaped with `\s`

**Comment by TheAssassin** [comment]
Interesting. Not that it'd excuse the failure, but why bother using spaces in paths at all?

I'll have to test that later again.

**Comment by develar** [comment]
> but why bother using spaces in paths at all?

1. I have no idea why AppImageLauncher decided to use `electron-webpack-quick-start 0.0.0` instead of `electron-webpack-quick-start` ;) Looks like a bug.
2. Even if you will fix bug "append version to app name with space", still, we cannot say to users "do not use space in your app title".

**Comment by TheAssassin** [comment]
> update: it seems AppImageLauncher uses file name as app name instead of name in a desktop file :(

Yes, that's the idea.

> Even if you will fix bug "append version to app name with space", still, we cannot say to users "do not use space in your app name".

Of course, as I stated in my last comment, this shall not be an excuse for broken behavior. Was just a bit curious why people auto-generate filenames with spaces, given all the issues they regularly produce.

**Comment by develar** [comment]
@TheAssassin Default artifact pattern for AppImage uses space.  `MyApp 1.2.1.AppImage` looks better (so vague and questionable ;)) on desktop, than `Tunepack-1.2.1.AppImage`.  But 

1. as now I don't see any strong reason why it was implemented so initially,
2. and AppImageLauncher in any case moves app to special place
3. and https://github.com/electron-userland/electron-builder/issues/1895#issuecomment-504987631

in electron-builder 21 it is changed.


```
    -         "path": "Test App ßW 1.1.0.AppImage",
    +         "path": "Test App ßW-1.1.0.AppImage",
```

**Comment by TheAssassin** [comment]
Let me clarify I don't insist on filenames being spaceless, that's really not the case here. Just was a bit curious there. Our tools also embed spaces in generated filenames if e.g., the desktop file contains spaces.

**Comment by TheAssassin** [comment]
Seems like it's the `Exec` key also needs to be wrapped in quotes to escape the spaces. I guess we need some actual escaping lib for that purpose. AFAIR `TryExec` does not need any special escaping, since it just contains a path, and no potential additional parameters. My experiments confirmed that. 

Therefore:
- GNOME 3 hides the entry not because of `TryExec` but because the path in `Exec` does not exist (it does not take the space as part of the path)
- we need some escaping/quoting for paths in `Exec` (quoting should probably be done by default, we need to check how to escape `"` characters, though)
- `TryExec` does not need any further actions.

@develar it would be really helpful if you could verify my findings. Just put quotes around the `Exec` and check if it works then.

**Comment by develar** [comment]
I have enough experience in this area, so, I can fully confirm, that:

1. `Exec` must be always escaped using `"` if contains characters like space. Otherwise you cannot run app. 
2. In `TryExec` space must be always escaped using `\s`. Otherwise you cannot run app.
3. Be aware about https://github.com/electron-userland/electron-builder/issues/2759#issuecomment-496942802 xdg-utils bug. So, you should escape `Exec` only if it   is really needed as workaround.

> Just put quotes around the Exec and check if it works then.

I already did it :) This change in electron-builder was driven by a fact, that I cannot figure out how to generate escaped `TryExec`  — https://github.com/AppImage/AppImageKit/issues/948#issuecomment-504523722

So, not clear what I should confirm — it is clear that right now AppImageLauncher can generate unescaped Exec and TryExec and it leads to a critical bug, that you don't see app in menu (because Desktop file is invalid) and cannot run it.

electron-builder desktop file always produces Exec=AppRun (and don't set TryExec anymore, because it is AppImageLauncher responsibility now).

**Comment by develar** [comment]
I can confirm that AppImageLauncher and electron-builder 21+ produced AppImage are ready for production use.

1. If you go to site and click on AppImage to download it, Firefox will show you Open Dialog like:

<img width="573" alt="Screenshot 2019-06-27 at 08 48 41" src="https://user-images.githubusercontent.com/350686/60243553-c6e7df80-98b8-11e9-90f6-3ca2a768c930.png">

    If you select "Open with", then AppImageLauncher dialog will be shown, and option "Integrate and run" works — app will be added to menu.

2. If you select "Save file", then AppImage file is downloaded to `~/Downloads`, and context menu contains correct "Open With AppImageLauncher". And click on file leads to correct AppImageLauncher dialog. 
<img width="743" alt="Screenshot 2019-06-27 at 08 54 31" src="https://user-images.githubusercontent.com/350686/60243822-5d1c0580-98b9-11e9-9492-414b4217f070.png">


So, thanks a lot of for AppImageLauncher, it works. You can close this ticket and open a new one about escaping spaces in Exec/TryExec.

I assume "Untrusted desktop entry file" related to weird Ubuntu bugs because of file permissions/whatever (when you copy file from macOS to Linux desktop). Regular users will download AppImage file from web.

**Comment by TheAssassin** [comment]
Well the core bug hasn't been resolved yet, we still don't escape things properly yet. We must fix that, but it's not as much of a prio, I guess.

I do not necessarily have to need to escape spaces in `TryExec=` by the way, but I guess it won't hurt either. Thanks for your

**Comment by azubieta** [comment]
Exec entries sapping is supported on xdg-utils-cxx see:
https://github.com/azubieta/xdg-utils/blob/3be709f8975438fc51e2c81ab549d8ce434d5d80/tests/DesktopEntry/TestDesktopEntryExecValue.cpp#L20-L25

But it's not used on AIL as far as I know.

**Comment by TheAssassin** [comment]
libappimage 1 will be used soon-ish, since I plan to make the change with version 2.0. See the milestone above.

**Comment by TheAssassin** [comment]
Should be resolved since #214.
